bladex/propeller.py

```diff
@@ -0,0 +1,71 @@
+"""
+Module for the propeller with shaft bottom-up parametrized construction.
+"""
+import os
+import numpy as np
+from bladex import Blade, Shaft
+import OCC.Core.TopoDS
+from OCC.Core.gp import gp_Dir, gp_Pnt, gp_Ax1, gp_Trsf
+from OCC.Core.IGESControl import IGESControl_Reader, IGESControl_Writer
+from OCC.Core.BRepBuilderAPI import BRepBuilderAPI_Transform, BRepBuilderAPI_Sewing
+from OCC.Core.BRepAlgoAPI import BRepAlgoAPI_Fuse
+from OCC.Extend.DataExchange import write_stl_file
+
+class Propeller(object):
+    """
+    Bottom-up parametrized propeller (including shaft) construction.
+
+    :param shaft.Shaft shaft: shaft to be added to the propeller
+    :param blade.Blade blade: blade of the propeller
+    :param int number_of blades: number of blades componing the propeller
+    """
+
+    def __init__(self, shaft, blade, number_of_blades):
+        self.shaft = shaft
+        self.blade = blade
+        self.number_of_blades = number_of_blades
+
+    def generate_propeller(self, propeller_and_shaft):
+        """
+        Generate and export the .iges and .stl CAD for the propeller with shaft.
+        This method requires PythonOCC (7.4.0) to be installed.
+
+        :raises RuntimeError: if the solid assembling of blades is not 
+            completed successfully
+        :param string propeller_and_shaft: path where to store the .iges and 
+            .stl CAD for the propeller and shaft
+        """
+        shaft_solid = self.shaft.generate_shaft_solid()
+        self.blade.apply_transformations(reflect=True)
+        blade_solid = self.blade.generate_blade_solid(max_deg=2,
+                                                 display=False,
+                                                 errors=None)
+        blades = []
+        blades.append(blade_solid)
+        for i in range(self.number_of_blades-1):
+            rot_dir = gp_Dir(1.0,0.0,0.0)
+            origin = gp_Pnt(0.0,0.0,0.0)        
+            rot_axis = gp_Ax1(origin, rot_dir)
+            transf = gp_Trsf()
+            transf.SetRotation(rot_axis, float(i+1)*2.0*np.pi/float(self.number_of_blades));
+            transformer = BRepBuilderAPI_Transform(blade_solid, transf);
+            blades.append(transformer.Shape())
```

Comment:
Not really clear to me why do not use the `rotate` method of the `Blade` class instead of manually transforming

Reply:
The rotate from blade.py is a method to rotate (in the (x,y)-plane) the planar sectional reference profile which is then used to build a blade. It seems to me that here the rotation is in the 3D space in order to correctly collocate the blades and form the propeller.

Reply:
The documentation (https://mathlab.github.io/BladeX/_summaries/bladex.blade.Blade.rotate.html#bladex.blade.Blade.rotate) of the function is not really consistent with your description:
```
        3D counter clockwise rotation about the X-axis of the Cartesian
        coordinate system, which is the axis of rotation of the propeller hub.
        The rotation matrix, :math:`R(\\theta)`, is used to perform rotation
        in the 3D Euclidean space about the X-axis, which is -- by default --
        the propeller axis of rotation.
```
Did I get it wrong?

Reply:
You can rotate a blade and then call `blade.generate_blade_solid`, for all the blades you need in order to compose the propeller

Reply:
Thank you for the suggestion.